modules/frontend/pipeline/async_weight_middleware.go

```diff
@@ -0,0 +1,103 @@
+package pipeline
+
+import (
+	"github.com/grafana/tempo/modules/frontend/combiner"
+	"github.com/grafana/tempo/pkg/traceql"
+)
+
+type RequestType int
+
+type WeightRequest interface {
+	SetWeight(int)
+	Weight() int
+}
+
+const (
+	Default RequestType = iota
+	TraceByID
+	TraceQLSearch
+	TraceQLMetrics
+)
+
+const (
+	DefaultWeight       int = 1
+	TraceQLSearchWeight int = 1
+	TraceByIDWeight     int = 2
+
+	maxTraceQLConditions int = 4
+	maxRegexConditions   int = 1
+)
+
+type weightRequestWare struct {
+	requestType RequestType
+	next        AsyncRoundTripper[combiner.PipelineResponse]
+}
+
+// It increments the weight of a retriyed request
+func IncrementRetriedRequestWeight(r WeightRequest) {
+	r.SetWeight(r.Weight() + 1)
+}
+
+// It returns a new weight request middleware
+func NewWeightRequestWare(rt RequestType) AsyncMiddleware[combiner.PipelineResponse] {
+	return AsyncMiddlewareFunc[combiner.PipelineResponse](func(next AsyncRoundTripper[combiner.PipelineResponse]) AsyncRoundTripper[combiner.PipelineResponse] {
+		return &weightRequestWare{
+			requestType: rt,
+			next:        next,
+		}
+	})
+}
+
+func (c weightRequestWare) RoundTrip(req Request) (Responses[combiner.PipelineResponse], error) {
+	c.setWeight(req)
+	return c.next.RoundTrip(req)
+}
+
+func (c weightRequestWare) setWeight(req Request) {
+	switch c.requestType {
+	case TraceByID:
+		req.SetWeight(TraceByIDWeight)
+	case TraceQLSearch, TraceQLMetrics:
+		setTraceQLWeight(req)
+	default:
+		req.SetWeight(DefaultWeight)
+	}
+}
+
+func setTraceQLWeight(req Request) {
+	var traceQLQuery string
+	query := req.HTTPRequest().URL.Query()
+	if query.Has("q") {
+		traceQLQuery = query.Get("q")
+	}
+	if query.Has("query") {
+		traceQLQuery = query.Get("query")
+	}
+	if traceQLQuery != "" {
```

Comment:
nit: for ease of reading i prefer.

```
if traceQLQuery == "" {
   req.SetWeight(TraceQLSearchWeight)
   return
}

...
```

this reduces nesting of the code below and very clearly communicates the logic taken when the query is not found